File: src/head_pose_detector.py
"""
Head Pose Detection Module
Uses MediaPipe Face Mesh to detect head orientation and track candidate's focus.
"""

import cv2
import mediapipe as mp
import numpy as np


class HeadPoseDetector:
    """
    Detects head pose using MediaPipe Face Mesh.
    Tracks head orientation to ensure candidate is looking at the screen.
    """
    
    def __init__(self):
        """Initialize MediaPipe Face Mesh for head pose detection."""
        self.mp_face_mesh = mp.solutions.face_mesh
        self.face_mesh = self.mp_face_mesh.FaceMesh(
            max_num_faces=1,
            refine_landmarks=True,
            min_detection_confidence=0.5,
            min_tracking_confidence=0.5
        )
        self.mp_drawing = mp.solutions.drawing_utils
        self.drawing_spec = self.mp_drawing.DrawingSpec(thickness=1, circle_radius=1)
        
        # Thresholds for direction detection (ratio-based)
        self.horizontal_threshold = 0.03  # For left/right
        self.vertical_threshold = 0.02    # For up/down
        
        # Current detected direction
        self.current_direction = "Forward"
        
    def detect_head_pose(self, frame):
        """
        Detect head pose from a video frame using nose position relative to face.
        
        Args:
            frame: Input BGR image frame
            
        Returns:
            tuple: (annotated_frame, pitch, yaw, roll, is_suspicious, direction)
        """
        # Convert BGR to RGB
        rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        results = self.face_mesh.process(rgb_frame)
        
        is_suspicious = False
        pitch, yaw, roll = 0, 0, 0
        
        if results.multi_face_landmarks:
            for face_landmarks in results.multi_face_landmarks:
                # Get image dimensions
                img_h, img_w, _ = frame.shape
                
                # Key landmark indices in MediaPipe Face Mesh:
                # 1 = Nose tip
                # 33 = Right eye outer corner
                # 263 = Left eye outer corner  
                # 61 = Right mouth corner
                # 291 = Left mouth corner
                # 199 = Chin bottom
                # 10 = Forehead top
                # 152 = Chin
                
                # Get nose tip position
                nose_tip = face_landmarks.landmark[1]
                nose_x = nose_tip.x
                nose_y = nose_tip.y
                
                # Get face boundaries using eye corners and chin
                left_eye = face_landmarks.landmark[263]  # Left eye outer
                right_eye = face_landmarks.landmark[33]   # Right eye outer
                chin = face_landmarks.landmark[152]
                forehead = face_landmarks.landmark[10]
                
                # Calculate face center (between eyes)
                face_center_x = (left_eye.x + right_eye.x) / 2
                face_center_y = (forehead.y + chin.y) / 2
                
                # Calculate face width and height for normalization
                face_width = abs(left_eye.x - right_eye.x)
                face_height = abs(chin.y - forehead.y)
                
                # Calculate normalized offset of nose from center
                horizontal_offset = (nose_x - face_center_x) / face_width if face_width > 0 else 0
                vertical_offset = (nose_y - face_center_y) / face_height if face_height > 0 else 0
                
                # Determine direction
                directions = []
                
                # Horizontal: positive = looking left, negative = looking right
                if horizontal_offset < -self.horizontal_threshold:
                    directions.append("Right")
                    is_suspicious = True
                elif horizontal_offset > self.horizontal_threshold:
                    directions.append("Left")
                    is_suspicious = True
                
                # Vertical: positive = looking down, negative = looking up
                if vertical_offset < -self.vertical_threshold:
                    directions.append("Up")
                    is_suspicious = True
                elif vertical_offset > self.vertical_threshold:
                    directions.append("Down")
                    is_suspicious = True
                
                if directions:
                    self.current_direction = " + ".join(directions)
                else:
                    self.current_direction = "Forward"
                
                # Set display color based on status
                if is_suspicious:
                    color = (0, 0, 255)  # Red
                    status = "LOOKING AWAY"
                else:
                    color = (0, 255, 0)  # Green
                    status = "LOOKING AT SCREEN"
                
                # Draw face mesh (optional - for debugging)
                # self.mp_drawing.draw_landmarks(frame, face_landmarks, self.mp_face_mesh.FACEMESH_CONTOURS)
                
                # Draw nose tip indicator
                nose_px = int(nose_x * img_w)
                nose_py = int(nose_y * img_h)
                cv2.circle(frame, (nose_px, nose_py), 5, color, -1)
                
                # Draw face center for reference
                center_px = int(face_center_x * img_w)
                center_py = int(face_center_y * img_h)
                cv2.circle(frame, (center_px, center_py), 5, (255, 255, 0), -1)
                
                # Draw line from center to nose
                cv2.line(frame, (center_px, center_py), (nose_px, nose_py), color, 2)
                
                # Display direction prominently
                cv2.putText(frame, f"Direction: {self.current_direction}", (20, 50), 
                           cv2.FONT_HERSHEY_SIMPLEX, 0.9, color, 2)
                cv2.putText(frame, f"Status: {status}", (20, 90), 
                           cv2.FONT_HERSHEY_SIMPLEX, 0.7, color, 2)
                
        return frame, pitch, yaw, roll, is_suspicious, self.current_direction
    
    def release(self):
        """Release resources."""
        self.face_mesh.close()

File: src/head_pose_detector.rs
/// Head Pose Detection Module
/// Uses OpenCV face detection and facial landmarks to detect head orientation
use opencv::{
    core::{Mat, Point, Scalar, Size, Vector},
    imgproc,
    objdetect,
    prelude::*,
    types::VectorOfRect,
};
use anyhow::Result;

pub struct HeadPoseDetector {
    face_cascade: objdetect::CascadeClassifier,
    horizontal_threshold: f64,
    vertical_threshold: f64,
    current_direction: String,
}

impl HeadPoseDetector {
    /// Initialize the head pose detector
    pub fn new() -> Result<Self> {
        // Load Haar Cascade for face detection
        let face_cascade = objdetect::CascadeClassifier::new(
            "haarcascade_frontalface_default.xml"
        )?;

        Ok(Self {
            face_cascade,
            horizontal_threshold: 0.03,
            vertical_threshold: 0.02,
            current_direction: "Forward".to_string(),
        })
    }

    /// Detect head pose from a video frame
    /// 
    /// Returns: (pitch, yaw, roll, is_suspicious, direction)
    pub fn detect_head_pose(&mut self, frame: &mut Mat) -> Result<(f64, f64, f64, bool, String)> {
        let mut gray = Mat::default();
        imgproc::cvt_color(frame, &mut gray, imgproc::COLOR_BGR2GRAY, 0)?;

        let mut faces = VectorOfRect::new();
        self.face_cascade.detect_multi_scale(
            &gray,
            &mut faces,
            1.3,
            5,
            0,
            Size::new(30, 30),
            Size::default(),
        )?;

        let mut is_suspicious = false;
        let pitch = 0.0;
        let yaw = 0.0;
        let roll = 0.0;

        if faces.len() > 0 {
            let face = faces.get(0)?;
            
            // Calculate face center
            let face_center_x = face.x + face.width / 2;
            let face_center_y = face.y + face.height / 2;

            // Get frame dimensions
            let frame_width = frame.cols() as i32;
            let frame_height = frame.rows() as i32;
            let frame_center_x = frame_width / 2;
            let frame_center_y = frame_height / 2;

            // Calculate normalized offsets
            let horizontal_offset = ((face_center_x - frame_center_x) as f64) / (face.width as f64);
            let vertical_offset = ((face_center_y - frame_center_y) as f64) / (face.height as f64);

            // Determine direction
            let mut directions = Vec::new();

            // Horizontal detection
            if horizontal_offset < -self.horizontal_threshold {
                directions.push("Right");
                is_suspicious = true;
            } else if horizontal_offset > self.horizontal_threshold {
                directions.push("Left");
                is_suspicious = true;
            }

            // Vertical detection
            if vertical_offset < -self.vertical_threshold {
                directions.push("Up");
                is_suspicious = true;
            } else if vertical_offset > self.vertical_threshold {
                directions.push("Down");
                is_suspicious = true;
            }

            self.current_direction = if directions.is_empty() {
                "Forward".to_string()
            } else {
                directions.join(" + ")
            };

            // Draw visualizations
            let color = if is_suspicious {
                Scalar::new(0.0, 0.0, 255.0, 0.0) // Red
            } else {
                Scalar::new(0.0, 255.0, 0.0, 0.0) // Green
            };

            // Draw face center
            imgproc::circle(
                frame,
                Point::new(face_center_x, face_center_y),
                5,
                color,
                -1,
                imgproc::LINE_8,
                0,
            )?;

            // Draw frame center
            imgproc::circle(
                frame,
                Point::new(frame_center_x, frame_center_y),
                5,
                Scalar::new(255.0, 255.0, 0.0, 0.0),
                -1,
                imgproc::LINE_8,
                0,
            )?;

            // Draw line from frame center to face center
            imgproc::line(
                frame,
                Point::new(frame_center_x, frame_center_y),
                Point::new(face_center_x, face_center_y),
                color,
                2,
                imgproc::LINE_8,
                0,
            )?;

            // Draw text
            let status = if is_suspicious {
                "LOOKING AWAY"
            } else {
                "LOOKING AT SCREEN"
            };

            imgproc::put_text(
                frame,
                &format!("Direction: {}", self.current_direction),
                Point::new(20, 50),
                imgproc::FONT_HERSHEY_SIMPLEX,
                0.9,
                color,
                2,
                imgproc::LINE_8,
                false,
            )?;

            imgproc::put_text(
                frame,
                &format!("Status: {}", status),
                Point::new(20, 90),
                imgproc::FONT_HERSHEY_SIMPLEX,
                0.7,
                color,
                2,
                imgproc::LINE_8,
                false,
            )?;
        }

        Ok((pitch, yaw, roll, is_suspicious, self.current_direction.clone()))
    }

    pub fn release(&self) {
        // Resources are automatically released in Rust
    }
}
